### PR TITLE
Improve documentation/comments on the random walk example

### PR DESCRIPTION
**Pull request by alan-cooney**
Makes this example more readable:

- Add some context to the README
- Add docstrings to the functions
- Add typings & fix all typing issues
- Give variables descriptive names
- Add thorough comments

### CODE REVIEW

File: examples/randomwalks/README.md [hunk 1]
```diff
@@ -1,14 +1,33 @@
...
+Paths are represented as strings of letters, with each letter corresponding to a
+node in the graph.
+
+## Training
+
+![Graph Example](graph-example.png)
+Source: Decision Transformer (Lili Chen et al. 2021)
+
+Paths are represented as strings of letters, with each letter corresponding to a
+node in a graph.
```
Comment (jon-tow) [inline examples/randomwalks/README.md:27]: Delete this it's a duplicate of the sentence above the `## Training` header

File: examples/randomwalks/ppo_randomwalks.py [hunk 2]
```diff
@@ -18,10 +18,14 @@ def main(hparams={}):
 
     trlx.train(
         "CarperAI/randomwalks",
-        reward_fn=lambda samples, **kwargs: metric_fn(samples)["optimality"],
+        # An "optimality" reward function is used, with scores in [0,1]
+        # depending on how close the path is to the shortest possible path.
+        reward_fn=lambda samples, p, o: metric_fn(samples, p, o)["optimality"],
```
Comment (jon-tow) [inline examples/randomwalks/ppo_randomwalks.py:23]: Revert this. RL Trainer calls the `reward_fn` with the keyword args `samples, prompts, outputs` and this breaks the call.

File: examples/randomwalks/randomwalks.py [hunk 3]
```diff
@@ -1,106 +1,244 @@
...
-def randexclude(rng: np.random.RandomState, n: int, exclude: int) -> int:
+def generate_rand_int_excluding(
+    random_state: np.random.RandomState,
+    max: int,
+    exclude: int
+) -> int:
+    """Random integer generator, excluding a specific number
+
+    Args:
+        random_state: Numpy random state
```
Comment (jon-tow) [inline examples/randomwalks/randomwalks.py:16]: Rever naming. `rng` is a very popular var name for an `np.random.RandomState` instance.

File: examples/randomwalks/README.md [hunk 4]
```diff
@@ -1,14 +1,33 @@
...
-https://github.com/kzl/decision-transformer/issues/48
+# Random Walks: Decision Tree Example
+
+This example uses the Toy Problem described in [Decision Transformer (Lili Chen
+et al. 2021)](https://arxiv.org/abs/2106.01345).
+
+## Game Description
+
+The task is to find the shortest path on a directed graph. The reward is based
+on tow optimal the path is compared to the shortest possible (bounded in [0,
```
Comment (jon-tow) [inline examples/randomwalks/README.md:9]: Fix misspelling:
> The reward is based on ~~tow~~ how

File: examples/randomwalks/randomwalks.py [hunk 5]
```diff
@@ -1,106 +1,244 @@
...
+            # If there is no path, use the maximum length instead
+            shortest_node_to_goal_lengths.append(max_length)
+
+    # Convert the shortest lengths list to a Tensor
+    shortest_lengths = torch.tensor(shortest_node_to_goal_lengths)
+
+    def metric_fn(
+        samples: List[str],
+        _prompts: Optional[List[str]] = None,
+        _outputs: Optional[List[str]] = None
```
Comment (jon-tow) [inline examples/randomwalks/randomwalks.py:169]: Revert these, they are unused.

File: examples/randomwalks/randomwalks.py [hunk 6]
```diff
@@ -1,106 +1,244 @@
...
+            max_length,
+            lengths
+        ).abs()
+        optimal_lengths = torch.as_tensor(sample_optimal_lengths)
 
         return {
-            "lengths": lengths,
-            # percentage-optimal \in (0, 1) when compared to the shortest path
-            "optimality": (max_length - bound_lengths) / (max_length - ref_lengths),
+            "lengths": sample_lengths,
```
Comment (jon-tow) [inline examples/randomwalks/randomwalks.py:231]: Revert this to `lengths` instead of `sample_lengths` to avoid type errors (`sample_lengths` is `Long`)

**Review by jon-tow — CHANGES_REQUESTED** [review]
Thanks for putting this together @alan-cooney. I had some trouble running things locally. Can you fix the requested changes and provide a `wandb` report for both examples to ensure everything is working as expected? Thanks!

**Comment by alan-cooney** [comment]
Thanks for the quick review @jon-tow 

PPO results - https://wandb.ai/alancooney/trlx/runs/eo1vxg53
ILQL results - https://wandb.ai/alancooney/trlx/runs/onfn4je1

## Typings fix

By the way, I had to fix the typings in trlx/trlx.py so that they work with the approach you prefer here. This fix is needed in any case as the typings were incorrect (the metric function only takes samples - it's just the reward function that also takes prompts and outsputs):

https://github.com/CarperAI/trlx/blob/84a07113d8ced1497c5a7941a02f9c534bb4e185/trlx/trainer/accelerate_base_trainer.py#L357

However, I'm happy to move this fix to a different PR if you want to keep the commit history clean (it's a small change, but it doesn't really belong here).

File: trlx/trlx.py [hunk 7]
```diff
@@ -14,9 +14,7 @@ def train(
     dataset: Optional[Iterable[Tuple[str, float]]] = None,
     prompts: Optional[List[str]] = None,
     eval_prompts: Optional[List[str]] = None,
-    metric_fn: Optional[
-        Callable[[List[str], List[str], List[str]], Dict[str, List[float]]]
-    ] = None,
+    metric_fn: Optional[Callable[[List[str]], Dict[str, List[float]]]] = None,
```
Comment (jon-tow) [inline trlx/trlx.py:17]: Revert this type-hint because it's actually what we want haha

File: trlx/trlx.py [hunk 8]
```diff
@@ -37,7 +35,7 @@ def train(
             Giving a single string `s` for the sample is a shorthand for (`tokenizer.bos_token`, `s`)
         prompts (List[str]): Prompts to sample off from during online training
         eval_prompts (List[str]): Prompts to periodically validate training on
-        metric_fn (Optional[Callable[[List[str], List[str], List[str]], Dict[str, List[float]]]]):
+        metric_fn ( Optional[Callable[[List[str]], Dict[str, List[float]]]]):
```
Comment (jon-tow) [inline trlx/trlx.py:38]: Revert this docstring change for reasons outlined above

File: examples/randomwalks/ppo_randomwalks.py [hunk 9]
```diff
@@ -18,10 +18,14 @@ def main(hparams={}):
...
-        reward_fn=lambda samples, **kwargs: metric_fn(samples)["optimality"],
+        # An "optimality" reward function is used, with scores in [0,1]
+        # depending on how close the path is to the shortest possible path.
+        reward_fn=lambda samples, prompts, outputs: metric_fn(samples)["optimality"],
+        # The prompts are simply the first notes (represented as letters) to
+        # start from
         prompts=prompts,
         eval_prompts=prompts,
-        metric_fn=lambda samples, **kwargs: metric_fn(samples),
+        metric_fn=metric_fn,
```
Comment (jon-tow) [inline examples/randomwalks/ppo_randomwalks.py:28]: Revert to match the intended `metric_fn` signature

**Review by jon-tow — CHANGES_REQUESTED** [review]
> By the way, I had to fix the typings in trlx/trlx.py so that they work with the approach you prefer here. This fix is needed in any case, as the typings were incorrect (the metric function only takes samples - it's just the reward function that also takes prompts and outputs):
> 
> https://github.com/CarperAI/trlx/blob/84a07113d8ced1497c5a7941a02f9c534bb4e185/trlx/trainer/accelerate_base_trainer.py#L357
> 
> However, I'm happy to move this fix to a different PR if you want to keep the commit history clean (it's a small change, but it doesn't really belong here).

Oh huh; this slipped under the radar. Thanks for the find 🙏 `metric_fn` should match the `reward_fn` signature  and be called as 
```python
metrics = self.metric_fn(
  samples=str_samples,
  prompts=str_prompts,
  outputs=str_outputs,
)
```
on the line that you've highlighted. It's a small enough change that we can squeeze it into this PR. Let me know if that's alright with you! Once that's done we should be good to merge 👍

**Comment by alan-cooney** [comment]
Makes sense! All done, and I've checked the runs work as well:

PPO - https://wandb.ai/alancooney/trlx/runs/agtv55vb
ILQL - https://wandb.ai/alancooney/trlx/runs/feb6oky6

**Comment by reciprocated** [comment]
Have to say this is a great work! However it's rather peculiar that I cannot reproduce your PPO wandb run from this branch despite my run being identical as from the `main`. It seems like a dependency difference and if it's not too laborious can you make a run from the `main` as well on your side just to confirm the suspicion? Thanks!

https://wandb.ai/sorry/trlx/reports/random_walks_document-v-main--VmlldzozMzkyNjE1

**Comment by alan-cooney** [comment]
Sure, this seems plausible. I get the same results on main as on this branch, so I think it's all fine.

Main - https://wandb.ai/alancooney/trlx/runs/ixhfy261
This branch - https://wandb.ai/alancooney/trlx/runs/agtv55vb (from above)

As I understand it, CUDA + different hardware can cause different results with the same random seed. But in terms of environments, I'm using this docker image https://github.com/CarperAI/trlx/pull/196 and then `pip install`, if you want to reproduce it.

**Review by jon-tow — APPROVED** [review]
I can also reproduce the runs from `main`:
- PPO: https://api.wandb.ai/links/jon-tow/879xsiiw 
- ILQL: https://api.wandb.ai/links/jon-tow/jbgyq8yg

Thanks a bunch, @alan-cooney!
